core/search/nested_filter.cpp

```diff
@@ -0,0 +1,584 @@
+#include "utils/type_limits.hpp"
+
+namespace iresearch {
+template<template<typename> typename M>
+struct HasScoreHelper<M<NoopAggregator>> : std::false_type {};
+}  // namespace iresearch
+
+namespace {
+
+// 1 2 3 4 5 6 7
+//
+// c c c p c c p
+// const auto prev = parent->seek_previous(target - 1);
+// const auto firstChild = child->seek(prev + 1); // first target's child
+// return parent->seek(firstChild + 1)
+//
+//
+// p p c c c
+// p c c c p c c
+//
+
+// do {
+//   const auto parent = parent->seek(target)
+//   const auto firstChild = child->seek(parent + 1);
+//   return parent->seek_prev(firstChild - 1)
+//
+//   while (parent->next() && parent->value() < firstChild);
+//
+//   const auto next = parent->value();
+//   child->seek(parent + 1);
+//   if (child < next) {
+//     break;
+//   }
+//   target = next;
+// while (true);
+
+// FIXME(gnusi): need to figure out the previous parent before target
+
+// FIXME(gnusi):
+// - tests for MinMatcher
+// - add AvgMatcher
+// - implement backwards seek for columnstore
+
+using namespace irs;
+
+class NoneMatcher;
+
+template<typename Matcher>
+class ChildToParentJoin final : public doc_iterator, private Matcher {
+ public:
+  ChildToParentJoin(doc_iterator::ptr&& parent, doc_iterator::ptr&& child,
+                    Matcher&& matcher) noexcept
+      : Matcher{std::move(matcher)},
+        parent_{std::move(parent)},
+        child_{std::move(child)} {
+    assert(parent_);
+    assert(child_);
+
+    parent_doc_ = irs::get<irs::document>(*parent_);
+    child_doc_ = irs::get<irs::document>(*child_);
+
+    std::get<attribute_ptr<cost>>(attrs_) =
+        irs::get_mutable<cost>(child_.get());
+
+    if constexpr (HasScore_v<Matcher>) {
+      PrepareScore();
+    }
+  }
+
+  doc_id_t value() const noexcept override {
+    return std::get<document>(attrs_).value;
+  }
+
+  attribute* get_mutable(irs::type_info::type_id id) override {
+    return irs::get_mutable(attrs_, id);
+  }
+
+  doc_id_t seek(doc_id_t target) override {
+    auto& doc = std::get<document>(attrs_);
+
+    if (IRS_UNLIKELY(target <= doc.value)) {
+      return doc.value;
+    }
+
+    auto parent = parent_->seek(target);
+
+    if (doc_limits::eof(parent)) {
+      doc.value = doc_limits::eof();
+      return doc_limits::eof();
+    }
+
+    return SeekInternal(parent);
+  }
+
+  bool next() override {
+    if (IRS_LIKELY(!doc_limits::eof(parent_doc_->value))) {
+      return !doc_limits::eof(SeekInternal(parent_doc_->value));
+    }
+
+    std::get<document>(attrs_).value = doc_limits::eof();
+    return false;
+  }
+
+ private:
+  friend Matcher;
+
+  using Attributes = std::tuple<document, attribute_ptr<cost>, score>;
+
+  doc_id_t SeekInternal(doc_id_t parent) {
+    assert(!doc_limits::eof(parent));
+    auto& doc = std::get<document>(attrs_);
+
+    // First valid child
+    child_->seek(parent + 1);
+
+    do {
+      const auto child = child_doc_->value;
+      assert(child > parent_doc_->value);
```

Comment:
```suggestion
      assert(child >= parent_doc_->value);
```
Looks like original  assert  could be violated for NoneMatcher in the following way:
We have the following index with just 1 parent and 1 child document (and child matches the desired filter)
1   2
p   c

1. we call ByNestedFilter::seek(1)  
2. parent_->value is 1
3. called SeekInternal(1)
4. child_ is seeked to 2 so child_->value is 2
5. doc.value is set to 1
6. parent_->next() is called so parent_->value = eof
7. NoneMatcher::Accept is called
8. condition on line 236 is false as eof is greater than 2
9. child_ is seeked to eof e.g. child_->value is eof
10. cycle is repeated and here we have child = eof and parent = eof
BTW on next run cycle will exit and eof will be returned so in general this works as needed.